subtests/docker_cli/systemd/systemd.py

```diff
@@ -0,0 +1,86 @@
+(``p4321.service``) and a ``script p4321-server.py``. An image is built with
+this script added. Systemd starts a container using this image & the container
+writes current time to port ``4321``. From host, the socket is read and the
+value is checked to see if it is a digit. Finally, container is stopped
+through systemd and system is cleaned up.
+
+"""
+
+import shutil
+import time
+import socket
+from os.path import exists
+from os import unlink
+from autotest.client import utils
+from autotest.client.shared.utils import is_port_free
+from dockertest.subtest import Subtest
+from dockertest.dockercmd import DockerCmd
+from dockertest.images import DockerImages
+from dockertest.output import mustpass
+from dockertest.containers import DockerContainers
+
+
+class systemd(Subtest):
+
+    def initialize(self):
+        super(systemd, self).initialize()
+        self.stuff['systemd_touchstone'] = False
+        shutil.copyfile(self.bindir + '/p4321.service',
+                        self.config['sysd_unitf_dest'])
+        dkrcmd = DockerCmd(self, 'build', ['--force-rm -t',
+                                           self.config['name'], self.bindir])
+        mustpass(dkrcmd.execute())
+
+    def run_once(self):
+        host = 'localhost'
+        port = 4321
+        self.sysd_action('daemon-reload')
+        self.sysd_action('start', self.config['sysd_unit_file'])
+        while is_port_free(port, host):
+            time.sleep(0.1)
+        time_from_socket = self.read_socket(host, port)
+        self.failif(not time_from_socket.isdigit(),
+                    "Data received from container is non-numeric: '%s'"
+                    % time_from_socket)
+        self.sysd_action('stop', self.config['sysd_unit_file'])
+
+    def cleanup(self):
+        super(systemd, self).cleanup()
+        cntr_name = DockerContainers(self).get_unique_name()
+        DockerImages(self).clean_all([cntr_name, self.config['name']])
+        filepath = self.config['sysd_unitf_dest']
+        if exists(filepath):
+            unlink(filepath)
+        if self.stuff['systemd_touchstone']:
+            self.sysd_action('daemon-reload')
+
+    def sysd_action(self, action=None, sysd_unit=None):
+        self.stuff['systemd_touchstone'] = True
+        command = "systemctl " + action
+        if sysd_unit:
+            command += " " + sysd_unit
+        utils.run(command, ignore_status=False)
+
+    @staticmethod
+    def read_socket(host, port):
+        num_read_try = 10
+        while num_read_try:
+            # using timeout 13-sec as it's 1-sec longer than two DNS timeouts
+            s = socket.create_connection((host, port), 13)
+            s.settimeout(1.0)
+            try:
+                time_from_socket = s.recv(512).strip()
+                if time_from_socket:
+                    s.shutdown(socket.SHUT_RDWR)
+                    return time_from_socket
+                num_read_try -= 1
+            except socket.timeout:
+                pass
```

Comment:
It might be nice to have a fallthrough here, perhaps raising DockerTestError with a helpful message indicating that we never got a response despite N retries.